android/src/com/unciv/app/CustomFileLocationHelperAndroid.kt

```diff
@@ -0,0 +1,131 @@
+package com.unciv.app
+
+import android.app.Activity
+import android.content.Intent
+import android.net.Uri
+import android.provider.DocumentsContract
+import android.provider.OpenableColumns
+import androidx.annotation.GuardedBy
+import com.unciv.logic.CustomFileLocationHelper
+import com.unciv.logic.SuccessfulLoadResult
+import com.unciv.ui.crashhandling.postCrashHandlingRunnable
+
+class CustomFileLocationHelperAndroid(private val activity: Activity) : CustomFileLocationHelper {
+
+    @GuardedBy("this")
+    private val callbacks = mutableListOf<ActivityCallback>()
+    @GuardedBy("this")
+    private var curActivityRequestCode = 100
+
+    override fun saveGame(
+        gameData: String,
+        suggestedLocation: String,
+        saveCompleteCallback: ((String?, Exception?) -> Unit)?
+    ) {
+
+        val requestCode = createActivityCallback { uri ->
+            if (uri == null) {
+                saveCompleteCallback?.invoke(null, null)
+                return@createActivityCallback
+            }
+
+            try {
+                activity.contentResolver.openOutputStream(uri, "rwt")
+                    ?.writer()
+                    ?.use {
+                        it.write(gameData)
+                    }
+                saveCompleteCallback?.invoke(uri.toString(), null)
+            } catch (ex: Exception) {
+                saveCompleteCallback?.invoke(null, ex)
+            }
+        }
+
+        // When we loaded, we returned a "content://" URI as file location.
+        val uri = Uri.parse(suggestedLocation)
+        val fileName = if (uri.scheme == "content") {
+            val cursor = activity.contentResolver.query(uri, null, null, null, null)
+            cursor.use {
+                // we should have a direct URI to a file, so first is enough
+                if (it?.moveToFirst() == true) {
+                    it.getString(it.getColumnIndex(OpenableColumns.DISPLAY_NAME))
+                } else {
+                    ""
+                }
+            }
+        } else {
+            // if we didn't load, this is some file name entered by the user
+            suggestedLocation
+        }
+
+        Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
+            type = "application/json"
+            putExtra(Intent.EXTRA_TITLE, fileName)
+            if (uri.scheme == "content") {
+                putExtra(DocumentsContract.EXTRA_INITIAL_URI, uri)
+            }
+            activity.startActivityForResult(this, requestCode)
+        }
+    }
+
+
+    override fun loadGame(loadCompleteCallback: (SuccessfulLoadResult?, Exception?) -> Unit) {
+        val callbackIndex = createActivityCallback { uri ->
+            if (uri == null) {
+                loadCompleteCallback(null, null)
+                return@createActivityCallback
+            }
+
+            var gameData: String? = null
+            var exception: Exception? = null
```

Comment:
This whole logic pipeline also looks cross- platform

- get platform-specific input stream
- attempt to read
- return result depending on successfulness of attempt